### PR TITLE
feat: Automatically apply asset caching to img src URLs

### LINKED ISSUE #167: Automatically apply asset caching to `img` src URLs
This would use the work from #133 and preact's "options hooks" (specifically the vnode creation hook) to automatically rewrite things like src props on img elements to use the "hashed" version of static files.

This would be done only for asset URLs that we definitely know to be static assets (the server knows the paths of all assets in the `static/` directory).

### PR DESCRIPTION
**Pull request by sylc**
fix https://github.com/lucacasonato/fresh/issues/167

> This would be done only for asset URLs that we definitely know to be static assets (the server knows the paths of all assets in the static/ directory).

I stumble on this because the preact hook needs to run on the client when an island is revived. This would require to pass the list of asset to the client, maybe using the island props, which would be cumbersome. Instead the hook assume that if the src path is starting with `/` then it must be a static asset as this is what fresh also assume.  WDYT ?

### CODE REVIEW

**Review by lucacasonato — COMMENTED** [review]
Very nice! Tests are failing though.

**Comment by sylc** [comment]
fixed

**Comment by thegaryroberts** [comment]
Thanks for documenting this. Had to do a lot of digging across examples to work out where to put the favicon. Should probably put a placeholder one (the fresh lemon) in the initial app to make this a bit easier. But the docs will help in any case.

File: src/runtime/utils.ts
Comment (thegaryroberts) [inline src/runtime/utils.ts]: This should just be an early return, you can skip the next else and reduce the nesting

File: src/runtime/utils.ts
Comment (thegaryroberts) [inline src/runtime/utils.ts]: You already have the typed props constant available.

File: src/runtime/utils.ts
Comment (thegaryroberts) [inline src/runtime/utils.ts]: You should be able to get rid of the explicit any with this:
 
```
    if (props.dataset.noAutoHashing) return;
```
As per:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset

Comment (thegaryroberts) [reply src/runtime/utils.ts]: I've not managed to test this at runtime though, but it does stop the typescript error.

Comment (lucacasonato) [reply src/runtime/utils.ts]: That probably doesn't work, considering this is not a "real" `HTMLImageElement` on the server.

**Review by lucacasonato — COMMENTED** [review]
This looks good, thanks @sylc. I want to do some minor refactors to the `asset` function before landing this, to make this more robust.

File: tests/main_test.ts
Comment (lucacasonato) [inline tests/main_test.ts]: Can you move this to `runtime/utils_test.ts`?

File: src/runtime/utils.ts
Comment (lucacasonato) [inline src/runtime/utils.ts]: With the new `asset` changes, this can be simplified to `if (typeof props.src === "string")` I think.

**Comment by lucacasonato** [comment]
The srcset parsing worries me slightly. I'll have to find some more test cases to check that it is robust enough.

**Comment by sylc** [comment]
an image with a src defined as `src="user.png"`  (no leading `/`) will be correctly served by fresh from the static folder but the hook will not apply. I'm not sure if this is just fine for now or a bug or a feature (the `data-no-auto-hashing` could be removed to leverage this. but the difference between the 2 path is very subtle.)
it is ready for another look. thanks.

**Comment by lucacasonato** [comment]
> an image with a src defined as src="user.png" (no leading /) will be correctly served by fresh from the static folder but the hook will not apply.

I think that is fine. We can always improve this in the future. We should be as conservative as possible in the beginning so we don't introduce bugs.

**Comment by lucacasonato** [comment]
@sylc I have some edits in the docs page, that I unfortunately can not push to your fork. Could you enable the "Allow maintainers to push to branch" option on this PR?

**Comment by sylc** [comment]
@lucacasonato  Sorry, Is that what you are referring to? it is already ticked 
image

**Comment by lucacasonato** [comment]
Mhh, weird. I can't push. Must be something on my side. Sorry for the trouble, I'll try figure it out.

**Comment by lucacasonato** [comment]
Oh, seems like it was an issue with my configuration (something to do with `git-lfs`).

**Comment by lucacasonato** [comment]
I have updated the `assetSrcSet` function to be a little more robust.

~~My final concern is that the vnode hooks are not installed client side yet, and `` components rendered from islands will have different `src` paths server side and client side. We thus need to install this vnode hook on the client too.~~ Retracted, I just missed where you do this during review.

**Review by lucacasonato — APPROVED** [review]
LGTM! Thanks @sylc. As a follow up we should add a test that the asset mapping in islands also works correctly client side.

**Comment by sylc** [comment]
> As a follow up we should add a test that the asset mapping in islands also works correctly client side.

@lucacasonato I agree, I was thinking to do that after https://github.com/lucacasonato/fresh/pull/197 is merged so that I can leverage the islands test setup in that PR

**Pull request merged**
